2025 shared contacts und Ansprechpersonen als first-class citizens

sschoeling opened the pull request:
Erste Version Shared Contacts. In dieser Version sind die meisten Sachen testbar, aber es ist noch bei weitem nicht alles abgedeckt.

Modell:

Es gibt jetzt 2 Join-Tabellen `customer_contacts` und `vendor_contacts`, die die Ansprechpartner mit den Kunden/Lieferanten verbinden. Diese Tabellen enthalten auch das `main` Flag was den Hauptansprechpartner kennzeichnet.

Rose-Integration ist über einen Helper gekapselt weil Rose leider überhaupt nicht damit klar kommt, dass es Extraattribute an den Join-Tabellen gibt (in diesem Fall das `main` Flag).

- `SL::DB::Customer->contacts` existiert weiterhin, ist aber **nur** für Lesezugriff. Schreibzugriff dadrauf macht das `main` Flag kaputt, und leider bietet Rose keine Möglichkeit den Setter hier zu deaktivieren. (dito `Vendor`)
- `SL::DB::Customer->main_contact` ist ein Shortcut um den Hauptansprechpartner zu kriegen. (dito `Vendor`)
- `SL::DB::Contact->detach` ist rausgeflogen und ist jetzt in  `SL::DB::Customer->link_contact` und `SL::DB::Customer->detach_contact`, da wird auch das main Flag gesetzt. Beide sind idempotent (können also aufgerufen werden obwohl schon linked/detached), aber speichern jetzt sofort anstatt auf das nächste `->save(cascade => 1)` zu warten.
- Filterzugriff muss die Jointabelle explizit angeben: `SL::DB::Manager::Contact->get_all(require_objects => ['customers'], query => [ 'customers.id' => ...])` funktioniert, der umgedrehte Fall ist in `SL::DB::Helper::CustomerVendorContacts` dokumentiert.

Was funktioniert:

- In Kunde/Lieferant gibt es jetzt ein zusätzliches Multselect um Ansprechpartner an/auszumachen.
- Das Hauptansprechpartnerflag kann in den Kunde/Lieferanten-masken weiter gesetzt werden.
- Kunden/Lieferantensuche mit Filter auf Ansprechpartner/Hauptansprechpartner funktioniert
- Ansprechpartnerdetails in Belegmasken als Popup funktioniert.
- Ansprechpartnersuche nach Telefonnummer
- Ansprechpartnerwahl in Briefen (Letter)
- Liste der Ansprechpartner in alten Rechnungscontrollern (is.pl/ir.pl)
- TopQuickSeach::Contact Suche

Umgeschrieben aber ungetestet (und maybe kaputt):

- Löschen von Ansprechpartnern mit den existierenden Buttons in Kunde/Lieferantenmaske
- Das Linken von existierenden Ansprechpartnern passiert über ein multiselect2side, was aber nicht auf diese Verwendung gestylt ist. Das schaut sich Cem gerade in #656 an.

Was funktioniert gar nicht:

- CsvImport::[Contacts, Order, DeliveryOrder] basiert darauf die Sachen am Rose-Objekt zu registrieren für ein späteres save, siehe oben.

Was fehlt noch:

- cp_cv_id wird vom Upgradescript noch nicht entfernt
- changelog

sschoeling committed:
Toter Code: Common::retrieve_customer_or_vendors

Teil von customer_or_vendor_selection.js

Der Rest davon Wurde 2018 in 2eaa39366f97994db3548a9b139d35b16fbb0903
entfernt.

sschoeling committed:
Shared Contacts: Model Änderungen

sschoeling committed:
Shared Contacts: IS::customer_details/IS::vendor_details joins

sschoeling committed:
Shared Contacts: cp_main feature

sschoeling committed:
Shared Contacts: detach nach Customer/Vendor verschoben

sschoeling committed:
Shared Contacts: cp_cv_id, cp_main und detach aufräumen 1

sschoeling committed:
Shared Contacts: cp_cv_id update 2

- detach/link_contact machen jetzt direkt Datenbankänderungen

sschoeling committed:
Shared Contacts: is/ir get CONTACTS

sschoeling committed:
Shared Contacts: cp_main speicherung und frontend

sschoeling committed:
Shared Contacts: Multiselect linked contact

sschoeling committed:
Shared Contacts: TopQuickSearch Contacts gefixt

sschoeling committed:
Shared Contacts: debug statement

sschoeling committed:
Shared Contacts: read only wrapper für die RODBO relationships

sschoeling committed:
Shared Contacts: alle contacts wrapper in einen Helper ausgelagert

ausserdem Tests.

sschoeling committed:
Shared Contacts: controller fix bei neu Anlegen von Kunden und entfernen des letzten contact

sschoeling committed:
Shared Contacts: Ansprechpartner Nummernkreise

sschoeling committed:
Shared Contacts: Ansprechpartner löschen: korrekt aufräumen

sschoeling committed:
Shared Contacts: confirm bei Ansprechpartner löschen

sschoeling committed:
Shared Contacts: TopQuickSearch nach cp_number

kiviniklas committed:
Shared Contacts: cp_number in Tab anzeigen

sschoeling committed:
Shared Contacts: CsvImport

kiviniklas committed:
multiselect2side: Option, Buttons zum Verschieben aller Einträge zu verbergen

kiviniklas committed:
Shared Contacts: multiselect2side Buttons "alle Verschieben" verbergen

Wenn auf der linken Seite sehr viele Ansprechpartner stehen, ist ein
Klick auf einen dieser Buttons eine unnötiges Hindernis.

kiviniklas committed:
Shared Contacts: Höhe des multiselect2side verdoppelt

Commit:
Shared Contacts: CustomerVendorContacts fix Join query_args

Sonst schlägt der Abruf im Controller lieferantenseitig fehl.

kiviniklas committed:
Shared Contacts: Suche nach Zugewiesenen korrigiert

kiviniklas committed:
Shared Contacts: verständlichere Begriffe für Suche gewählt

kiviniklas committed:
Shared Contacts: locales fix

kiviniklas committed:
Contacts 1st class citizens: Controller und Picker

kiviniklas committed:
Contacts 1st class citizens: in Bericht integriert

kiviniklas committed:
Contacts 1st class citizens: Locales

kiviniklas committed:
Contacts 1st class citizens: in TopQuickSearch integriert

kiviniklas committed:
TopQuickSearch Contacts: Nach Person gruppieren, nicht nach Firma

kiviniklas committed:
S:Presenter:Tag input_phone_tag für CTI Actions auf Telefonnnummern

kiviniklas committed:
templates: input_phone_tag nutzen

kiviniklas commented:
# Ansprechpersonen als first-class citizens

## Weshalb ist das nötig?

Das Editieren der Ansprechpersonen innerhalb des CustomerVendor Controllers ist
- unhandlich zu bedienen
- inperformant, wenn man 10000 Ansprechpersonen in das Template lädt, um sie im Multiselect2Side darzustellen für die Zuordnung

Besser ist es,
- klickbare Links zu einem neuen Contacts Controller anzuzeigen, in dem eine Ansprechpersonen editieren, neu anlegen oder löschen lässt und
- wechselseitig über einen Ansprechpersonen-Picker (neu implementiert) bzw. Kunden-/Lieferantenpicker die Verknüpfung hinzuzfügen

## Wie sieht es aus?

a)
<img width="1334" height="737" alt="grafik" src="https://github.com/user-attachments/assets/1a91ee6d-c7c9-4076-afb8-3eba3b8d5f72" />

b)
<img width="1334" height="273" alt="grafik" src="https://github.com/user-attachments/assets/b0fba56c-f72f-4011-89d0-25e751ae884b" />


## TODO

- [x] Berechtigungsprüfung im Contacts Controller erledigt
- [x] Bereinigen von ContactDepartment und ContactTitle sollte nun die bei allen Ansprechpersonen gesetzten Werte berücksichtigen, der Bezug auf einen Customer/Vendor ist nicht mehr sinnvoll

## Weitere Anmerkungen

- Bei den Telefonnummern wurde das Hörer-Symbol bei aktiviertem CTI nicht im Browser angezeigt. Das ist nun auf einen Presenter umgestellt, damit an die E-Mail-Felder angeglichen, und funktioniert wieder.

kiviniklas commented:
<del>Exkurs für später:
Wir können noch überlegen, aus dem booleschen main Flag eine Sortierreihenfolge zu machen mit der Konvention: der erste Eintrag ist der Primäre / was zuvor main war.
Das würde gut zur join table passen, denn die Einträge werden ja ohne hin sortiert. Damit wäre die Reihenfolge der Einträge im contacts Array am Customer/Vendor Rose Objekt explizit festgelegt. Von der API her finde ich das schöner, als etwas gekünstelt mit dem main Flag zu arbeiten.
Aus SL::DB::Customer->main_contact würde einfach SL::DB::Customer->contacts->[0].</del>

<del>Grundsätzlich benötigen wir für Kundenanpassungen Attribute an der join Tabelle, d.h. die bisherige Arbeit wird eh benötigt.
</del> verworfen. Das main Flag explizit über ein Eingabefeld zu steuern, gefällt Moritz und Sven besser - der Vorschlag, das über die Sortierung zu handhaben, ist für Anwender zu implizit.

kiviniklas committed:
Contacts 1st class citizens: Locales

kiviniklas committed:
Contacts 1st class citizens: in Menü integriert

kiviniklas committed:
Contacts 1st class citizens: Berechtigungen cleanup

kiviniklas committed:
Contacts 1st class citizens: ungenutzte Titles und Departments bereinigen

kiviniklas committed:
Contacts 1st class citizens: aus K/L anlegen in neuem Tab

... damit ungespeicherte Änderungene im CustomerVendor Controller erhalten
bleiben.

kiviniklas committed:
Contacts 1st class citizens: Locales

kiviniklas committed:
Contacts 1st class citizens: aus K/L anlegen per Workflow

kiviniklas committed:
Contacts 1st class citizens: Locales

kiviniklas committed:
Contacts 1st class citizens: DB Änderungen erst bei Action "Speichern" ausführen

kiviniklas committed:
Contacts 1st class citizens: JS nutze commit_one zum Hinzufügen

kiviniklas committed:
Contacts 1st class citizens: Controller-Actions per JS auslösen

Entspricht mehr der Konvention der neueren Controller.
Erlaubt Validieren und macht es, falls nötig, einfacher flash_later() im
Contoller zu verwenden.

kiviniklas committed:
Contacts 1st class citizens: fix Presenter

kiviniklas committed:
ContactsPicker: angeglichen an PartPicker

kiviniklas committed:
ContactPicker: Test page hinzugefügt

kiviniklas committed:
ContactPicker: Test page: Locales

kiviniklas committed:
ContactPicker: fix pre filled picker

kiviniklas commented:
Angepasst werden musste noch der Trigger in sql/Pg-upgrade2/cleanup_after_customer_vendor_deletion.sql:
`-- @description: Nach Löschen von Kunden/Lieferanten via Trigger auch Ansprechpersonen/Lieferadressen löschen`
(ist erledigt)

kiviniklas committed:
Shared Contacts: DB Trigger cleanup_after_customer_vendor_deletion fix

kiviniklas commented:
wird abgelöst von: #791
